### PR TITLE
add basic algebra operations for MeshDLRImFreq Gf and density functions

### PR DESCRIPTION
**Pull request by the-hampel**
* The basic operations += / -= / *= where missing for MeshDLRImFreq Gf. This commit adds those, wrappes them to Python, and adds tests.
* add density function for DLRImFreq and DLRImTime

### CODE REVIEW

File: c++/triqs/gfs/functions/density.hpp [hunk 1]
```diff
@@ -60,13 +60,11 @@ namespace triqs {
 
     auto density(MemoryGf<mesh::dlr_imtime> auto const &g) {
       // FIXME : workaround for static_assert(false). Fix on clang >= 17
-      static_assert(sizeof(g) == 0, "density(gf<dlr_imtime, ...>) is not supported. Use a gf<dlr, ...>.");
-      return 0;
+      return density(make_gf_dlr(g));
```
Comment (Wentzell) [inline c++/triqs/gfs/functions/density.hpp:62]: Remove redundant comment in line above.

File: c++/triqs/gfs/functions/density.hpp [hunk 2]
```diff
@@ -60,13 +60,11 @@ namespace triqs {
...
     }
     auto density(MemoryGf<mesh::dlr_imfreq> auto const &g) {
       // FIXME : workaround for static_assert(false). Fix on clang >= 17
-      static_assert(sizeof(g) == 0, "density(gf<dlr_imfreq, ...>) is not supported. Use a gf<dlr, ...>.");
-      return 0;
+      return density(make_gf_dlr(g));
```
Comment (Wentzell) [inline c++/triqs/gfs/functions/density.hpp:65]: Remove redundant comment in line above.

File: c++/triqs/gfs/legacy_for_python_api.hpp [hunk 3]
```diff
@@ -86,5 +86,24 @@ namespace triqs {
...
+    inline void operator+=(gf_view<dlr_imfreq> g, std::complex<double> a) { operator+=(g, make_matrix(a * nda::eye<double>(g.target_shape()[0]))); }
+    inline void operator-=(gf_view<dlr_imfreq> g, std::complex<double> a) { operator-=(g, make_matrix(a * nda::eye<double>(g.target_shape()[0]))); }
+
+    // Same for scalar valued
+    // THIS MUST be kept for python operations
+    // specific operations (for legacy python code).
+    inline void operator+=(gf_view<dlr_imfreq, scalar_valued> g, std::complex<double> a) { g.data() += a; }
+
+    inline void operator-=(gf_view<dlr_imfreq, scalar_valued> g, std::complex<double> a) { g.data() -= a; }
+
```
Comment (Wentzell) [inline c++/triqs/gfs/legacy_for_python_api.hpp]: The following should cover all the cases implemented above and generalize it for other Mesh types. Please check. Also `make_matrix` may then be redundant.

```suggestion
    // basic MeshDLRImFreq operations
    template <Mesh M, nda::MemoryMatrix Mat>
    inline void operator+=(gf_view<M> g, Mat const &mat) {
      for (auto mp: g.mesh()) g[mp] += mat;
    }

    template <Mesh M, nda::MemoryMatrix Mat>
    inline void operator-=(gf_view<M> g, Mat const &mat) {
      for (auto mp: g.mesh()) g[mp] -= mat;
    }

    template <Mesh M>
    inline void operator+=(gf_view<M> g, dcomplex a) { g += make_regular(a * nda::eye<double>(g.target_shape()[0])); }
    
    template <Mesh M>
    inline void operator-=(gf_view<M> g, dcomplex a) { g -= make_regular(a * nda::eye<double>(g.target_shape()[0])); }

    // Same for scalar valued
    // THIS MUST be kept for python operations
    // specific operations (for legacy python code).
    template <Mesh M>
    inline void operator+=(gf_view<M, scalar_valued> g, dcomplex a) { g.data() += a; }

    template <Mesh M>
    inline void operator-=(gf_view<M, scalar_valued> g, dcomplex a) { g.data() -= a; }

```

File: python/triqs/gf/wrapped_aux_desc.py [hunk 4]
```diff
@@ -72,6 +72,7 @@ def all_calls():
                    'retime' : 'ReTime',
                    'legendre' : 'Legendre',
                    'dlr'    : 'DLR',
+                   'dlr_imfreq' : 'DLRImFreq',
```
Comment (Wentzell) [inline python/triqs/gf/wrapped_aux_desc.py:76]:
```suggestion
                   'dlr_imtime' : 'DLRImTime',
                   'dlr_imfreq' : 'DLRImFreq',
```

File: python/triqs/gf/wrapped_aux_desc.py [hunk 5]
```diff
@@ -146,8 +147,8 @@ def all_calls():
 # invert auxiliary tool
 m.add_function("void _gf_invert_data_in_place(array_view <dcomplex, 3> a)", doc = "Aux function for inversion")
 
-# For legacy Python code : authorize g + Matrix
-for M in ['imfreq', 'refreq'] :
+# For legacy Python code : authorize g + Matrix functions, which are defined in legacy_for_python_api.hpp
+for M in ['imfreq', 'refreq', 'dlr_imfreq']:
```
Comment (Wentzell) [inline python/triqs/gf/wrapped_aux_desc.py]:
```suggestion
for M in ['imfreq', 'imtime', 'refreq', 'retime', 'brzone', 'cyclat', 'legendre', 'dlr', 'dlr_imfreq', 'dlr_imtime']:
```

**Review by Wentzell — CHANGES_REQUESTED** [review]
Thank you @the-hampel for these improvement! See comments / suggestions below.

File: c++/triqs/gfs/functions/density.hpp (hunk 1 above)
Comment (the-hampel) [reply c++/triqs/gfs/functions/density.hpp:62]: removed in 9ca05d61c3402e2c051011c3a51aee22674ef40d

File: c++/triqs/gfs/functions/density.hpp (hunk 2 above)
Comment (the-hampel) [reply c++/triqs/gfs/functions/density.hpp:65]: 9ca05d61c3402e2c051011c3a51aee22674ef40d

File: c++/triqs/gfs/legacy_for_python_api.hpp (hunk 3 above)
Comment (the-hampel) [reply c++/triqs/gfs/legacy_for_python_api.hpp]: In the comment here I wrote that this works for MeshDLRImFreq objects, but I guess this is more general now. What includes MemoryMatrix objects? And how can we maybe make in the comment clear what this does? I had a hard time to find this code.

File: test/python/base/gf_base_op.py [hunk 6]
```diff
@@ -144,6 +152,25 @@ def check_pickle(g):
...
+        # some basic checks for MeshImTime
+        tau_mesh = MeshImTime(beta=self.beta, S = "Fermion", n_tau = 1000)
+
+        ga_tau = Gf(mesh=tau_mesh, target_shape=(2,2), name = "a1Block")
+        gb_tau = Gf(mesh=tau_mesh, target_shape=(2,2), name = "b1Block")
+
+        G_tau = BlockGf(name_list = ('a','b'), block_list = (ga_tau,gb_tau), make_copies = False)
+        for tau in G_tau.mesh:
+            for block, gf in G_tau:
+                gf[tau] = onefermion(tau, 1.413, 1e-10)
```
Comment (Wentzell) [inline test/python/base/gf_base_op.py]:
```suggestion
        for block, g in G_tau:
            for tau in g.mesh:
                g[tau] = onefermion(tau, 1.413, 1e-10)
```

File: test/python/base/gf_base_op.py [hunk 7]
```diff
@@ -144,6 +152,25 @@ def check_pickle(g):
...
+
+        # Arithmetic operations
+        G2_tau = G_tau.copy()
+        G2_tau << 1.5 * G_tau
+
+        G2_tau += G_tau
+        G2_tau['b'] += np.eye(G_tau['a'].data.shape[0])
```
Comment (Wentzell) [inline test/python/base/gf_base_op.py]:
```suggestion
        G2_tau['b'] += np.eye(G_tau['b'].target_shape[0])
```

**Review by Wentzell — CHANGES_REQUESTED** [review]
Thank you @the-hampel for the updates.
Can be merged after addressing the comments below.

**Review by Wentzell — APPROVED** [review]

**Pull request merged**